File: api/src/test/java/org/pmiops/workbench/actionaudit/targetproperties/TargetPropertyExtractorTest.java
package org.pmiops.workbench.actionaudit.targetproperties;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.pmiops.workbench.access.AccessTierService;
import org.pmiops.workbench.model.ResearchPurpose;
import org.pmiops.workbench.model.SpecificPopulationEnum;
import org.pmiops.workbench.model.User;
import org.pmiops.workbench.model.Workspace;

class TargetPropertyExtractorTest {
  private Workspace workspace;

  @BeforeEach
  void setUp() {
    long now = System.currentTimeMillis();
    User creator = new User();
    creator.setEmail("user@fake-research-aou.org");

    var researchPurpose1 =
        new ResearchPurpose()
            .intendedStudy("stubbed toes")
            .additionalNotes("I really like the cloud.")
            .approved(true)
            .ancestry(true)
            .anticipatedFindings("I expect to find a lot of stubbed toes.")
            .commercialPurpose(true)
            .controlSet(true)
            .diseaseFocusedResearch(true)
            .diseaseOfFocus("toes")
            .drugDevelopment(true)
            .educational(true)
            .intendedStudy("stubbed toes")
            .methodsDevelopment(true)
            .otherPopulationDetails("Something about toes I guess")
            .populationDetails(
                List.of(
                    SpecificPopulationEnum.DISABILITY_STATUS,
                    SpecificPopulationEnum.SEXUAL_ORIENTATION,
                    SpecificPopulationEnum.SEXUAL_ORIENTATION))
            .populationHealth(true)
            .reasonForAllOfUs("I like it.")
            .reviewRequested(true)
            .socialBehavioral(true)
            .timeRequested(now)
            .timeReviewed(now);

    workspace =
        new Workspace()
            .etag("etag_1")
            .name("DbWorkspace 1")
            .terraName("dbworkspace1")
            .namespace("aou-rw-local1-c4be869a")
            .cdrVersionId("1")
            .creator(creator)
            .accessTierShortName(AccessTierService.REGISTERED_TIER_SHORT_NAME)
            .researchPurpose(researchPurpose1);
  }

  @Test
  void testGetsWorkspaceProperties() {
    var propertyValuesByName =
        ModelBackedTargetProperty.getPropertyValuesByName(
            WorkspaceTargetProperty.values(), workspace);
    assertThat(propertyValuesByName.get(WorkspaceTargetProperty.NAME.getPropertyName()))
        .isEqualTo("DbWorkspace 1");
    assertThat(propertyValuesByName).hasSize(WorkspaceTargetProperty.values().length);
  }
}

File: api/src/test/java/org/pmiops/workbench/actionaudit/targetproperties/TargetPropertyExtractorTest.kt
package org.pmiops.workbench.actionaudit.targetproperties

import com.google.common.truth.Truth.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.pmiops.workbench.access.AccessTierService
import org.pmiops.workbench.model.ResearchPurpose
import org.pmiops.workbench.model.Workspace

class TargetPropertyExtractorTest {
    private var workspace: Workspace? = null
    private var researchPurpose1: ResearchPurpose? = null

    @BeforeEach
    fun setUp() {
        researchPurpose1 =
            ResearchPurpose()
                .apply { intendedStudy = "stubbed toes" }
                .apply { additionalNotes = "I really like the cloud." }

        val now = System.currentTimeMillis()

        workspace =
            Workspace()
                .apply { name = "DbWorkspace 1" }
                .apply { id = "fc-id-1" }
                .apply { namespace = "aou-rw-local1-c4be869a" }
                .apply { creator = "user@fake-research-aou.org" }
                .apply { cdrVersionId = "1" }
                .apply { researchPurpose = researchPurpose1 }
                .apply { creationTime = now }
                .apply { lastModifiedTime = now }
                .apply { etag = "etag_1" }
                .apply { accessTierShortName = AccessTierService.REGISTERED_TIER_SHORT_NAME }
                .published(false)
    }

    @Test fun testGetsWorkspaceProperties() {
        val propertyValuesByName =
            TargetPropertyExtractor.getPropertyValuesByName(
                WorkspaceTargetProperty.values(),
                workspace!!,
            )
        assertThat(propertyValuesByName[WorkspaceTargetProperty.NAME.propertyName])
            .isEqualTo("DbWorkspace 1")
        assertThat(propertyValuesByName).hasSize(19)
    }

    @Test fun testGetTargetPropertyEnumByTargetClass() {
        val result = TargetPropertyExtractor.getTargetPropertyEnumByTargetClass()
        assertThat(result).hasSize(2)
        assertThat(result[Workspace::class.java]!!).isEqualTo(WorkspaceTargetProperty::class.java)
    }

    @Test fun testGetPropertyEnum() {
        val result = TargetPropertyExtractor.getTargetPropertyEnum(Workspace::class.java)
        assertThat(result).isEqualTo(WorkspaceTargetProperty::class.java)
    }
}
